File: typescript/day18_part1_2020.ts
import * as fs from 'fs';

const evaluate = (expr: string): number => {
    const tokens = expr.match(/\d+|[\+\*\(\)]/g) || [];
    const output: string[] = [];
    const operators: string[] = [];

    const precedence = (op: string): number => (op === '+' || op === '*') ? 1 : 0;

    const applyOperator = (op: string, b: number, a: number): number => {
        return op === '+' ? a + b : a * b;
    };

    const evaluateStack = () => {
        const b = Number(output.pop());
        const a = Number(output.pop());
        const op = operators.pop()!;
        output.push(applyOperator(op, b, a).toString());
    };

    for (const token of tokens) {
        if (!isNaN(Number(token))) {
            output.push(token);
        } else if (token === '(') {
            operators.push(token);
        } else if (token === ')') {
            while (operators.length && operators[operators.length - 1] !== '(') {
                evaluateStack();
            }
            operators.pop(); // Remove '('
        } else {
            while (operators.length && precedence(operators[operators.length - 1]) >= precedence(token)) {
                evaluateStack();
            }
            operators.push(token);
        }
    }

    while (operators.length) {
        evaluateStack();
    }

    return Number(output[0]);
};

const main = () => {
    const input = fs.readFileSync('input.txt', 'utf-8').trim().split('\n');
    const total = input.reduce((sum, line) => sum + evaluate(line), 0);
    console.log(total);
};

main();

File: elixir/day18_part1_2020.exs

defmodule Main do
  def main do
    sum =
      File.stream!("input.txt")
      |> Stream.map(&String.trim/1)
      |> Stream.filter(&(&1 != ""))
      |> Stream.map(&evaluate/1)
      |> Enum.sum()

    IO.puts(sum)
  end

  defp evaluate(expr) do
    tokens = tokenize(expr)
    {_, [result]} = eval_tokens(tokens, {[], []})
    result
  end

  defp tokenize(expr) do
    expr
    |> String.replace("(", "( ")
    |> String.replace(")", " )")
    |> String.split(~r/\s+/, trim: true)
  end

  defp eval_tokens([], {ops, vals}) do
    {ops, vals} = apply_remaining_ops(ops, vals)
    {ops, vals}
  end

  defp eval_tokens([token | rest], {ops, vals}) do
    case token do
      "(" ->
        eval_tokens(rest, {[token | ops], vals})

      "+" ->
        {ops, vals} = apply_until_parenthesis(ops, vals)
        eval_tokens(rest, {[token | ops], vals})

      "*" ->
        {ops, vals} = apply_until_parenthesis(ops, vals)
        eval_tokens(rest, {[token | ops], vals})

      ")" ->
        {ops, vals} = apply_until_parenthesis(ops, vals)
        eval_tokens(rest, {tl(ops), vals})

      _ ->
        value = String.to_integer(token)
        eval_tokens(rest, {ops, [value | vals]})
    end
  end

  defp apply_until_parenthesis(ops, vals) do
    case ops do
      [] -> {ops, vals}
      ["(" | _] -> {ops, vals}
      [op | rest_ops] ->
        [a, b | rest_vals] = vals
        result = apply_op(op, b, a)
        apply_until_parenthesis(rest_ops, [result | rest_vals])
    end
  end

  defp apply_remaining_ops(ops, vals) do
    case ops do
      [] -> {ops, vals}
      [op | rest_ops] ->
        [a, b | rest_vals] = vals
        result = apply_op(op, b, a)
        apply_remaining_ops(rest_ops, [result | rest_vals])
    end
  end

  defp apply_op("+", a, b), do: a + b
  defp apply_op("*", a, b), do: a * b
end

Main.main()
